File: js(ts)-solutions/2467MostProfitablePathInATree.ts
function mostProfitablePath(edges: number[][], bob: number, amount: number[]): number {
    let graph = new Map(), bobTimer = new Map();
    let bobPath: number[] = [];

    // function to build the tree
    const buildGraph = () => {
        for (const [a, b] of edges) {
            if (!graph.has(a)) graph.set(a, []);
            if (!graph.has(b)) graph.set(b, []);

            graph.get(a).push(b);
            graph.get(b).push(a);
        }
    }

    // function to find bob's path to node 0
    const findPathToRoot = (node: number, seen: Set<number>): boolean => {
        bobPath.push(node);
        seen.add(node);

        if (node === 0) return true;

        for (const adjacentNode of graph.get(node)) {
            if (!(seen.has(adjacentNode)) && findPathToRoot(adjacentNode, seen)) {
                return true;
            }
        }

        bobPath.pop();
        return false;
    }

    // function to check who got to this node first
    const whoReachedFirst = (node: number, time: number): number => {
        if (!bobTimer.has(node)) {
            return amount[node];
        } else if (bobTimer.get(node) === time) {
            return amount[node] / 2;
        } else if (bobTimer.get(node) < time) {
            return 0
        } else {
            return amount[node]
        }
    }

    // function to find alice's max income by travelling to the optimal leaf node
    const getMaxIncome = (node: number, time: number, seen: Set<number>): number => {
        // if a leaf node has been found
        if (node !== 0 && graph.get(node).length === 1) return whoReachedFirst(node, time);

        const currIncome = whoReachedFirst(node, time);

        seen.add(node);
        let maxIncome = -Infinity;
        for (const adjacentNode of graph.get(node)) {
            if (!seen.has(adjacentNode)) {
                maxIncome = Math.max(maxIncome, getMaxIncome(adjacentNode, time + 1, seen));
            }
        }

        return currIncome + maxIncome;
    }

    buildGraph();
    findPathToRoot(bob, new Set())

    // store the time bob reaches each node in his path
    bobPath.forEach((node, index) => {
        bobTimer.set(node, index);
    })

    const maxIncome = getMaxIncome(0, 0, new Set());
    return maxIncome;
};

File: leetcode-questions/2467-most-profitable-path-in-a-tree/2467-most-profitable-path-in-a-tree.py
class Solution:
    def __init__(self):
        self.graph = {}
        self.amount = []
        self.bob_path = []

    def buildGraph(self, edges):
        for a, b in edges:
            if b not in self.graph:
                self.graph[b] = []
            if a not in self.graph:
                self.graph[a] = []

            self.graph[b].append(a)
            self.graph[a].append(b)

    def findPathToRoot(self, node, seen):
        self.bob_path.append(node)
        seen.add(node)

        if node == 0:
            return True

        for adjacent_node in self.graph[node]:
            if adjacent_node not in seen and self.findPathToRoot(adjacent_node, seen):
                return True

        seen.remove(node)
        self.bob_path.pop()
        return False

    def traverse(self, node, time, seen):
        # if reached a leaf node, return
        if len(self.graph[node]) == 1 and node != 0:
            # this would mean bob got to this node first
            if node not in self.tracker or (
                node in self.tracker and self.tracker[node] > time
            ):
                return self.amount[node]
            elif self.tracker[node] == time:
                return self.amount[node] / 2
            return 0

        curr_income = 0
        if node not in self.tracker or (
            node in self.tracker and self.tracker[node] > time
        ):
            curr_income = self.amount[node]
        elif self.tracker[node] == time:
            curr_income = int(self.amount[node] / 2)

        seen.add(node)
        max_income = -float("inf")
        for neighbor in self.graph[node]:
            if neighbor not in seen:
                max_income = max(max_income, self.traverse(
                    neighbor, time + 1, seen))
        seen.remove(node)

        return max_income + curr_income

    def mostProfitablePath(self, edges: List[List[int]], bob: int, amount: List[int]) -> int:
        self.amount = amount
        self.buildGraph(edges)

        if self.findPathToRoot(bob, set()):
            self.tracker = {node: time for time,
                            node in enumerate(self.bob_path)}

        return self.traverse(0, 0, set())
